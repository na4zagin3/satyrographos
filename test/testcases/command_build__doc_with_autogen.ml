module StdList = List

open Satyrographos_testlib
open TestLib

open Shexp_process

let satyristes =
{|
(lang "0.0.3")

(doc
  (name "example-doc")
  (build
    ((satysfi "doc-example.saty" "-o" "doc-example-ja.pdf")
     (make "build-doc")))
  (dependencies (localized-today))
  (autogen ($libraries)))
|}

let satysfi_grcnum_opam =
  "satysfi-grcnum.opam", TestLib.opam_file_for_test
    ~name:"satysfi-grcnum"
    ~version:"0.1"
    ()

let makefile =
{|
PHONY: build-doc
build-doc:
	@echo "Target: build-doc"
|}

let lockdown_yaml =
  "lockdown.yaml", {|satyrographos: 0.0.3
dependencies:
- Opam
- packages:
  - name: ocaml
    version: 4.09.0
  - name: satyrographos
    version: 0.0.2.7
  - name: satysfi
    version: 0.0.5+dev2020.09.05
  repos:
  - name: default
    url: https://opam.ocaml.org/
  - name: satysfi-external
    url: git+https://github.com/gfngfn/satysfi-external-repo.git
autogen:
  '$today':
    datetime: 2020-11-06T00:46:35.000000+09:00
    tzname: Asia/Tokyo
|}

let files =
  [
    satysfi_grcnum_opam;
    "Satyristes", satyristes;
    "README.md", "@@README.md@@";
    "doc-grcnum.saty", "@@doc-grcnum.saty@@";
    "doc-example.saty", "@@doc-example.saty@@";
    "Makefile", makefile;
    lockdown_yaml;
  ]

let env ~dest_dir:_ ~temp_dir : Satyrographos.Environment.t t =
  let open Shexp_process.Infix in
  let pkg_dir = FilePath.concat temp_dir "pkg" in
  let prepare_pkg =
    PrepareDist.empty pkg_dir
    >> prepare_files pkg_dir files
  in
  let empty_dist = FilePath.concat temp_dir "empty_dist" in
  let prepare_dist = PrepareDist.empty empty_dist in
  let opam_reg = FilePath.concat temp_dir "opam_reg" in
  let log_file = exec_log_file_path temp_dir in
  let prepare_opam_reg =
    PrepareOpamReg.(prepare opam_reg localizedTodayFiles)
    >> PrepareOpamReg.(prepare opam_reg baseFiles)
  in
  let bin = FilePath.concat temp_dir "bin" in
  prepare_pkg
  >> prepare_dist
  >> prepare_opam_reg
  >> PrepareBin.prepare_bin bin log_file
  >>| read_env ~opam_reg ~dist_library_dir:empty_dist

let () =
  let verbose = false in
  let main env ~dest_dir:_ ~temp_dir ~outf =
    let names = Some ["example-doc"] in
    (* let dest_dir = FilePath.concat dest_dir "dest" in *)
    Satyrographos_command.Build.build_command
      ~outf
      ~verbose
      ~buildscript_path:(FilePath.concat temp_dir "pkg/Satyristes")
      ~build_dir:(FilePath.concat temp_dir "pkg/_build" |> Option.some)
      ~env
      ~names
  in
  let post_dump_dirs ~dest_dir:_ ~temp_dir =
    let pkg_dir = FilePath.concat temp_dir "pkg" in
    [pkg_dir]
  in
  eval (test_install ~post_dump_dirs env main)
