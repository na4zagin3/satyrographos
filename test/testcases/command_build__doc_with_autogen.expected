Installing packages
------------------------------------------------------------
= Pin projects

= Build modules: (example-doc)

== Build module example-doc
== Install dependencies: (localized-today)
=== Build docs
Target: build-doc
================
Reading runtime dist: @@temp_dir@@/empty_dist
Reading opam libraries: (base localized-today)
Not gathering system fonts
Generating autogen libraries
Generating autogen library $today
autogen:$today: Using lockdowned values
Generating autogen library $libraries
Installing libraries: ($libraries $today dist localized-today)
Removing destination @@temp_dir@@/pkg/_build/satysfi/dist
Installation completed!
------------------------------------------------------------
@@temp_dir@@/pkg
@@temp_dir@@/pkg/Makefile
@@temp_dir@@/pkg/README.md
@@temp_dir@@/pkg/Satyristes
@@temp_dir@@/pkg/_build
@@temp_dir@@/pkg/_build/satysfi
@@temp_dir@@/pkg/_build/satysfi/dist
@@temp_dir@@/pkg/_build/satysfi/dist/.satyrographos
@@temp_dir@@/pkg/_build/satysfi/dist/metadata
@@temp_dir@@/pkg/_build/satysfi/dist/packages
@@temp_dir@@/pkg/_build/satysfi/dist/packages/$libraries.satyg
@@temp_dir@@/pkg/_build/satysfi/dist/packages/$today.satyg
@@temp_dir@@/pkg/_build/satysfi/dist/packages/localized-today
@@temp_dir@@/pkg/_build/satysfi/dist/packages/localized-today/localized-today.satyh
@@temp_dir@@/pkg/doc-example-ja.pdf
@@temp_dir@@/pkg/doc-example.saty
@@temp_dir@@/pkg/doc-grcnum.saty
@@temp_dir@@/pkg/lockdown.yaml
@@temp_dir@@/pkg/satysfi-grcnum.opam
------------------------------------------------------------
diff -Nr @@empty_dir@@/Makefile @@temp_dir@@/pkg/Makefile
0a1,5
> 
> PHONY: build-doc
> build-doc:
> 	@echo "Target: build-doc"
> 
diff -Nr @@empty_dir@@/README.md @@temp_dir@@/pkg/README.md
0a1
> @@README.md@@
diff -Nr @@empty_dir@@/Satyristes @@temp_dir@@/pkg/Satyristes
0a1,11
> 
> (lang "0.0.3")
> 
> (doc
>   (name "example-doc")
>   (build
>     ((satysfi "doc-example.saty" "-o" "doc-example-ja.pdf")
>      (make "build-doc")))
>   (dependencies (localized-today))
>   (autogen ($libraries)))
> 
diff -Nr @@empty_dir@@/_build/satysfi/dist/metadata @@temp_dir@@/pkg/_build/satysfi/dist/metadata
0a1,2
> ((version 1) (libraryName $libraries) (libraryVersion 0.1) (compatibility ())
>  (dependencies ()) (autogen (($today ()))))
diff -Nr @@empty_dir@@/_build/satysfi/dist/packages/$libraries.satyg @@temp_dir@@/pkg/_build/satysfi/dist/packages/$libraries.satyg
0a1,15
> let _ =
>   (display-message) (#` [Warning] Satyrographos: Package $libraries is an experimental autogen package.`)
> let _ =
>   (display-message) (#` [Warning] Satyrographos: Its API is unstable; will thus be backward-incompatibly changed.`)
> let _ =
>   (display-message) (#` [Warning] Satyrographos: Furthermore, the package itself may be renamed or removed.`)
> type library = (| name : string; version : string; |)
> module Libraries : sig
>   val list : library list
> end = struct
>   let list =
>     [ (| name = `$today`; version = `0.1`; |);
>       (| name = `dist`; version = ` `; |);
>       (| name = `localized-today`; version = `0.1`; |); ]
> end
diff -Nr @@empty_dir@@/_build/satysfi/dist/packages/$today.satyg @@temp_dir@@/pkg/_build/satysfi/dist/packages/$today.satyg
0a1,13
> let _ =
>   (display-message) (#` [Warning] Satyrographos: Package $today is an experimental autogen package.`)
> let _ =
>   (display-message) (#` [Warning] Satyrographos: Its API is unstable; will thus be backward-incompatibly changed.`)
> let _ =
>   (display-message) (#` [Warning] Satyrographos: Furthermore, the package itself may be renamed or removed.`)
> module Today : sig
>   val datetime : string
>   val tzname : string
> end = struct
>   let datetime = `2020-11-06T00:46:35.000000+09:00`
>   let tzname = `Asia/Tokyo`
> end
diff -Nr @@empty_dir@@/_build/satysfi/dist/packages/localized-today/localized-today.satyh @@temp_dir@@/pkg/_build/satysfi/dist/packages/localized-today/localized-today.satyh
0a1
> @require: $today
diff -Nr @@empty_dir@@/doc-example-ja.pdf @@temp_dir@@/pkg/doc-example-ja.pdf
0a1
> @@doc-example.saty@@
diff -Nr @@empty_dir@@/doc-example.saty @@temp_dir@@/pkg/doc-example.saty
0a1
> @@doc-example.saty@@
diff -Nr @@empty_dir@@/doc-grcnum.saty @@temp_dir@@/pkg/doc-grcnum.saty
0a1
> @@doc-grcnum.saty@@
diff -Nr @@empty_dir@@/lockdown.yaml @@temp_dir@@/pkg/lockdown.yaml
0a1,20
> satyrographos: 0.0.3
> dependencies:
> - Opam
> - packages:
>   - name: ocaml
>     version: 4.09.0
>   - name: satyrographos
>     version: 0.0.2.7
>   - name: satysfi
>     version: 0.0.5+dev2020.09.05
>   repos:
>   - name: default
>     url: https://opam.ocaml.org/
>   - name: satysfi-external
>     url: git+https://github.com/gfngfn/satysfi-external-repo.git
> autogen:
>   '$today':
>     datetime: 2020-11-06T00:46:35.000000+09:00
>     tzname: Asia/Tokyo
> 
diff -Nr @@empty_dir@@/satysfi-grcnum.opam @@temp_dir@@/pkg/satysfi-grcnum.opam
0a1,30
> opam-version: "2.0"
> synopsis: "Test Package"
> name: "satysfi-grcnum"
> version: "0.1"
> description: """
> Test package for SATySFi
> """
> maintainer: "SAKAMOTO Noriaki <mrty.ityt.pt@gmail.com>"
> authors: "SAKAMOTO Noriaki <mrty.ityt.pt@gmail.com>"
> license: "LGPL-3.0-or-later"
> homepage: "https://github.com/na4zagin3/satysfi-fss"
> dev-repo: "git+https://github.com/na4zagin3/satysfi-fss.git"
> bug-reports: "https://github.com/na4zagin3/satysfi-fss/issues"
> depends: [
> 
>   "satysfi" {>= "0.0.5" & < "0.0.6"}
>   "satyrographos" {>= "0.0.2.6" & < "0.0.3"}
> 
>   "satysfi-base" {>= "1.3.0" & < "2"}
>   "satysfi-fonts-junicode" {>= "1" & < "2"}
> 
> ]
> build: [ ]
> install: [
>   ["satyrographos" "opam" "install"
>    "--name" "test-package"
>    "--prefix" "%{prefix}%"
>    "--script" "%{build}%/Satyristes"]
> ]
> 
------------------------------------------------------------
Command invoked:
opam install --yes satysfi-localized-today
Command invoked:
satysfi --version
Command invoked:
satysfi -C @@temp_dir@@/pkg/_build/satysfi --no-default-config doc-example.saty -o doc-example-ja.pdf
doc-example.saty -> doc-example-ja.pdf
